bittensor/utils/version.py

```diff
@@ -0,0 +1,74 @@
+from typing import Optional
+from pathlib import Path
+import time
+from packaging.version import Version
+
+import bittensor
+import requests
+
+VERSION_CHECK_THRESHOLD = 86400
+
+
+def _get_version_file_path() -> Path:
+    return Path.home() / ".bittensor" / ".last_known_version"
+
+
+def _get_version_from_file(version_file: Path) -> Optional[str]:
+    if not version_file.exists():
+        bittensor.logging.debug("No bitensor version file found")
+        return None
+
+    mtime = version_file.stat().st_mtime
+    bittensor.logging.debug(f"Found version file, last modified: {mtime}")
+    diff = time.time() - mtime
+
+    if diff >= VERSION_CHECK_THRESHOLD:
+        bittensor.logging.debug("Version file expired")
+        return None
+
+    try:
+        return version_file.read_text()
+    except OSError:
+        bittensor.logging.exception("Failed to read version file")
+        return None
+
+
+def _get_version_from_pypi(timeout: int = 15) -> str:
+    bittensor.logging.debug(
+        f"Checking latest Bittensor version at: {bittensor.__pipaddress__}"
+    )
+    try:
+        response = requests.get(bittensor.__pipaddress__, timeout=timeout)
+        latest_version = response.json()["info"]["version"]
+        return latest_version
+    except requests.exceptions.RequestException:
+        bittensor.logging.exception("Failed to get latest version from pypi")
+        raise
+
+
+def get_and_save_latest_version(timeout: int = 15) -> str:
```

Comment:
I ended up just renaming the function. Extracting the writing part seemed unnatural because the version file is used in both parts. The resulting code is still not ideal, @agoncharov-reef maybe you have better ideas on structuring this?